### ISSUE TITLE
Update to ESPHome 4.0 does not work and still 3.4.1

### ISSUE THREAD

**Issue by BananaJon** [Blackymas/NSPanel_HA_Blueprint#1086]
I already updated my NSPanel in the past. But this time I have issue with the update through ESPHome. 

I used this HowTo
https://github.com/Blackymas/NSPanel_HA_Blueprint/wiki/(DE)-HowTo---Alle-wichtigen-Dinge-die-man-wissen-sollte#2-update-blueprint

1. I updated the blueprint by replacing the nspanel_blueprint.yaml. I deleted the old file and uploaded the new one. After that I restarted HomeAssistant
![image](https://github.com/Blackymas/NSPanel_HA_Blueprint/assets/76597486/b184e2ba-3153-49f3-a4d1-b0b4104037b4)

2. In ESPHome I validated the configuration and when the green hook comes up I installied the config. 
![image](https://github.com/Blackymas/NSPanel_HA_Blueprint/assets/76597486/6a706f5e-02f6-45f9-94b8-1e2a799991f5)

3. I did not check if the update was successful because there was no obvious error message. I then directly made the TFT update to version 4.0 and that was also successful. 

I do not know why the update to ESPHome 4.0 does not work. I already tried to Clean Build Files and the installed it again and I have already tried it via the web panel as an OTA upload. 

![image](https://github.com/Blackymas/NSPanel_HA_Blueprint/assets/76597486/5bc585e8-31db-4ef8-84d3-d3eb827ddace)

![image](https://github.com/Blackymas/NSPanel_HA_Blueprint/assets/76597486/593af777-396b-485b-8e3d-a2f7b8eba9df)

In legacy mode it says Update succesfully. 

This is a part of the log during installation:

```
INFO Successfully compiled program.
INFO Resolving IP address of nspanel01.local
INFO  -> 172.26.20.50
INFO Uploading /data/nspanel01/.pioenvs/nspanel01/firmware.bin (1171280 bytes)
Uploading: [============================================================] 100% Done...

INFO Waiting for result...
INFO OTA successful
INFO Successfully uploaded program.
INFO Starting log output from nspanel01.local using esphome API
```

**Comment by edwardtfn** [comment]
What is not working?

Can you see anything in you panel's screen? Which versions you see at boot page? (TFT, ESPHome and Blueprint)

Have you tried updating via Wireless instead of saving files? What happens?

**Comment by BananaJon** [comment]
At the boot page it shows:
TFT 4.0 
ESPHome 3.4.1
Blueprint nothing bzw. Retry

First via wireless update and then OTA via NSPanel WebConsole . What is meant with instead of saving files. I only know wireless and manually.  

I still have the following three error messages in HomeAssistant. Is it related to this or is it because the update is not yet complete? 

![image](https://github.com/Blackymas/NSPanel_HA_Blueprint/assets/76597486/6a38c6eb-e8d0-4ce4-8bfb-47307d09d3f1)

![image](https://github.com/Blackymas/NSPanel_HA_Blueprint/assets/76597486/7a1bfc88-d191-4666-a4fb-d47a7958ad2c)

![image](https://github.com/Blackymas/NSPanel_HA_Blueprint/assets/76597486/2c79048b-3456-4065-88e0-de6e97937de2)

**Comment by edwardtfn** [comment]
> I still have the following three error messages in HomeAssistant. Is it related to this or is it because the update is not yet complete?


Those all look related to ESPHome not updated.

**Comment by edwardtfn** [comment]
Looks like TFT and Blueprint are both in the right version, so it's ESPHome only...

What happens when you try to install wirelessly? Could you please share the logs with the error?

**Comment by BananaJon** [comment]
INFO Reading configuration /config/esphome/nspanel01.yaml...
INFO Updating https://github.com/Blackymas/NSPanel_HA_Blueprint@main
INFO Reverting changes to https://github.com/Blackymas/NSPanel_HA_Blueprint@main -> 9aa8421117b7b76c352710f84d2947966293b764
WARNING GPIO4 is a Strapping PIN and should be avoided.
Attaching external pullup/down resistors to strapping pins can cause unexpected failures.
See https://esphome.io/guides/faq.html#why-am-i-getting-a-warning-about-strapping-pins
INFO Generating C++ source...
INFO Compiling app...
Processing nspanel01 (board: esp32dev; framework: arduino; platform: platformio/espressif32 @ 5.2.0)
--------------------------------------------------------------------------------
HARDWARE: ESP32 240MHz, 320KB RAM, 4MB Flash
 - toolchain-xtensa-esp32 @ 8.4.0+2021r2-patch3
Dependency Graph
|-- AsyncTCP-esphome @ 1.2.2
|-- WiFi @ 2.0.0
|-- FS @ 2.0.0
|-- Update @ 2.0.0
|-- ESPAsyncWebServer-esphome @ 2.1.0
|   |-- AsyncTCP-esphome @ 1.2.2
|-- DNSServer @ 2.0.0
|-- ESPmDNS @ 2.0.0
|-- ArduinoJson @ 6.18.5
|-- WiFiClientSecure @ 2.0.0
|-- HTTPClient @ 2.0.0
Compiling /data/nspanel01/.pioenvs/nspanel01/src/main.cpp.o
Linking /data/nspanel01/.pioenvs/nspanel01/firmware.elf
RAM:   [=         ]  13.4% (used 44040 bytes from 327680 bytes)
Flash: [======    ]  63.5% (used 1165513 bytes from 1835008 bytes)
Building /data/nspanel01/.pioenvs/nspanel01/firmware.bin
Creating esp32 image...
Successfully created esp32 image.
esp32_create_combined_bin(["/data/nspanel01/.pioenvs/nspanel01/firmware.bin"], ["/data/nspanel01/.pioenvs/nspanel01/firmware.elf"])
Wrote 0x12df50 bytes to file /data/nspanel01/.pioenvs/nspanel01/firmware-factory.bin, ready to flash to offset 0x0
========================= [SUCCESS] Took 21.66 seconds =========================
INFO Successfully compiled program.
INFO Resolving IP address of nspanel01.local
INFO  -> 172.26.20.50
INFO Uploading /data/nspanel01/.pioenvs/nspanel01/firmware.bin (1171280 bytes)
Uploading: [============================================================] 100% Done...

INFO Waiting for result...
INFO OTA successful
INFO Successfully uploaded program.
INFO Starting log output from nspanel01.local using esphome API
INFO Successfully connected to nspanel01.local
[16:25:08][I][app:102]: ESPHome version 2023.4.0-dev compiled on Sep 23 2023, 16:24:26
[16:25:08][C][wifi:504]: WiFi:
[16:25:08][C][wifi:362]:   Local MAC: 80:64:6F:XX:XX:XX
[16:25:08][C][wifi:363]:   SSID: 'XXXXXXXXX'[redacted]
[16:25:08][C][wifi:364]:   IP Address: 172.26.20.50
[16:25:08][C][wifi:366]:   BSSID: 4A:8F:5A:XX:XX:XX[redacted]
[16:25:08][C][wifi:367]:   Hostname: 'nspanel01'
[16:25:08][C][wifi:369]:   Signal strength: -70 dB ▂▄▆█
[16:25:08][C][wifi:373]:   Channel: 1
[16:25:08][C][wifi:374]:   Subnet: 255.255.255.0
[16:25:08][C][wifi:375]:   Gateway: 172.26.20.254
[16:25:08][C][wifi:376]:   DNS1: 172.26.20.254
[16:25:08][C][wifi:377]:   DNS2: 0.0.0.0
[16:25:08][C][logger:293]: Logger:
[16:25:08][C][logger:294]:   Level: DEBUG
[16:25:08][C][logger:295]:   Log Baud Rate: 0
[16:25:08][C][logger:296]:   Hardware UART: UART0
[16:25:08][C][uart.arduino_esp32:108]: UART Bus 1:
[16:25:08][C][uart.arduino_esp32:109]:   TX Pin: GPIO16
[16:25:08][C][uart.arduino_esp32:110]:   RX Pin: GPIO17
[16:25:08][C][uart.arduino_esp32:112]:   RX Buffer Size: 256
[16:25:08][C][uart.arduino_esp32:114]:   Baud Rate: 115200 baud
[16:25:08][C][uart.arduino_esp32:115]:   Data Bits: 8
[16:25:08][C][uart.arduino_esp32:116]:   Parity: NONE
[16:25:08][C][uart.arduino_esp32:117]:   Stop bits: 1
[16:25:08][C][ledc.output:160]: LEDC Output:
[16:25:08][C][ledc.output:161]:   Pin GPIO21
[16:25:08][C][ledc.output:162]:   LEDC Channel: 0
[16:25:08][C][ledc.output:163]:   PWM Frequency: 1000.0 Hz
[16:25:08][C][ledc.output:164]:   Bit depth: 16
[16:25:08][C][template.binary_sensor:018]: Template Binary Sensor 'nspanel01 Nextion display'
[16:25:08][C][template.binary_sensor:018]:   Device Class: 'connectivity'
[16:25:08][C][uptime.sensor:031]: Uptime Sensor 'nspanel01 uptime'
[16:25:08][C][uptime.sensor:031]:   Device Class: 'duration'
[16:25:08][C][uptime.sensor:031]:   State Class: 'total_increasing'
[16:25:08][C][uptime.sensor:031]:   Unit of Measurement: 's'
[16:25:08][C][uptime.sensor:031]:   Accuracy Decimals: 0
[16:25:08][C][uptime.sensor:031]:   Icon: 'mdi:timer-outline'
[16:25:08][C][template.text_sensor:021]: Template Sensor 'nspanel01 Notification Label'
[16:25:08][C][template.text_sensor:021]: Template Sensor 'nspanel01 Notification Text'
[16:25:08][C][template.text_sensor:021]: Template Sensor 'nspanel01 Settings Entity'
[16:25:08][C][template.switch:075]: Template Switch 'nspanel01 Notification unread'
[16:25:08][C][template.switch:098]:   Restore Mode: restore defaults to OFF
[16:25:08][C][template.switch:060]:   Restore State: NO
[16:25:08][C][template.switch:061]:   Optimistic: YES
[16:25:08][C][template.switch:075]: Template Switch 'nspanel01 Notification sound'
[16:25:08][C][template.switch:098]:   Restore Mode: restore defaults to OFF
[16:25:08][C][template.switch:060]:   Restore State: NO
[16:25:08][C][template.switch:061]:   Optimistic: YES
[16:25:08][C][template.switch:075]: Template Switch 'nspanel01 Confirmation Message'
[16:25:08][C][template.switch:098]:   Restore Mode: restore defaults to OFF
[16:25:08][C][template.switch:060]:   Restore State: NO
[16:25:08][C][template.switch:061]:   Optimistic: YES
[16:25:08][C][switch.gpio:075]: GPIO Switch 'nspanel01 Relay 1'
[16:25:08][C][switch.gpio:098]:   Restore Mode: restore defaults to OFF
[16:25:08][C][switch.gpio:031]:   Pin: GPIO22
[16:25:08][D][nspanel:949]: no page-jump
[16:25:08][C][switch.gpio:075]: GPIO Switch 'nspanel01 Relay 2'
[16:25:08][C][switch.gpio:098]:   Restore Mode: restore defaults to OFF
[16:25:08][C][switch.gpio:031]:   Pin: GPIO19
[16:25:08][C][switch.gpio:075]: GPIO Switch 'nspanel01 Screen Power'
[16:25:08][C][switch.gpio:098]:   Restore Mode: always ON
[16:25:08][C][switch.gpio:031]:   Pin: GPIO4
[16:25:08][C][template.switch:075]: Template Switch 'nspanel01 Sleep mode'
[16:25:08][C][template.switch:086]:   Device Class: 'switch'
[16:25:08][C][template.switch:098]:   Restore Mode: restore defaults to OFF
[16:25:08][C][template.switch:060]:   Restore State: NO
[16:25:08][C][template.switch:061]:   Optimistic: NO
[16:25:08][C][template.switch:075]: Template Switch 'nspanel01 Relay 1 Local Fallback'
[16:25:08][C][template.switch:098]:   Restore Mode: restore defaults to OFF
[16:25:08][C][template.switch:060]:   Restore State: NO
[16:25:08][C][template.switch:061]:   Optimistic: YES
[16:25:08][C][template.switch:075]: Template Switch 'nspanel01 Relay 2 Local Fallback'
[16:25:08][C][template.switch:098]:   Restore Mode: restore defaults to OFF
[16:25:08][C][template.switch:060]:   Restore State: NO
[16:25:08][C][template.switch:061]:   Optimistic: YES
[16:25:08][C][template.number:050]: Template Number 'nspanel01 Display Brightness'
[16:25:08][C][template.number:050]:   Unit of Measurement: '%'
[16:25:08][C][template.number:051]:   Optimistic: YES
[16:25:08][C][template.number:052]:   Update Interval: 60.0s
[16:25:08][C][template.number:050]: Template Number 'nspanel01 Display Brightness Dimdown'
[16:25:08][C][template.number:050]:   Unit of Measurement: '%'
[16:25:08][C][template.number:051]:   Optimistic: YES
[16:25:08][C][template.number:052]:   Update Interval: 60.0s
[16:25:08][C][template.number:050]: Template Number 'nspanel01 Temperature Correction'
[16:25:08][C][template.number:050]:   Unit of Measurement: '°C'
[16:25:08][C][template.number:051]:   Optimistic: YES
[16:25:08][C][template.number:052]:   Update Interval: 60.0s
[16:25:08][C][template.number:050]: Template Number 'nspanel01 Page Timeout'
[16:25:08][C][template.number:051]:   Optimistic: YES
[16:25:08][C][template.number:052]:   Update Interval: 60.0s
[16:25:09][C][gpio.binary_sensor:015]: GPIO Binary Sensor 'nspanel01 Left Button'
[16:25:09][C][gpio.binary_sensor:016]:   Pin: GPIO14
[16:25:09][C][gpio.binary_sensor:015]: GPIO Binary Sensor 'nspanel01 Right Button'
[16:25:09][C][gpio.binary_sensor:016]:   Pin: GPIO27
[16:25:09][C][rtttl:018]: Rtttl
[16:25:09][C][restart.button:017]: Restart Button 'nspanel01 Restart'
[16:25:09][C][status:034]: Status Binary Sensor 'nspanel01 Status'
[16:25:09][C][status:034]:   Device Class: 'connectivity'
[16:25:09][C][adc:087]: ADC Sensor 'ntc_source'
[16:25:09][C][adc:087]:   Device Class: 'voltage'
[16:25:09][C][adc:087]:   State Class: 'measurement'
[16:25:09][C][adc:087]:   Unit of Measurement: 'V'
[16:25:09][C][adc:087]:   Accuracy Decimals: 2
[16:25:09][C][adc:097]:   Pin: GPIO38
[16:25:09][C][adc:112]:  Attenuation: 11db
[16:25:09][C][adc:126]:   Update Interval: 60.0s
[16:25:09][C][resistance:010]: Resistance Sensor 'resistance_sensor'
[16:25:09][C][resistance:010]:   State Class: 'measurement'
[16:25:09][C][resistance:010]:   Unit of Measurement: 'Ω'
[16:25:09][C][resistance:010]:   Accuracy Decimals: 1
[16:25:09][C][resistance:010]:   Icon: 'mdi:flash'
[16:25:09][C][resistance:011]:   Configuration: DOWNSTREAM
[16:25:09][C][resistance:012]:   Resistor: 11200.00Ω
[16:25:09][C][resistance:013]:   Reference Voltage: 3.3V
[16:25:09][C][version.text_sensor:021]: Version Text Sensor 'nspanel01 ESPhome Version'
[16:25:09][C][version.text_sensor:021]:   Icon: 'mdi:new-box'
[16:25:09][C][nextion:117]: Nextion:
[16:25:09][C][nextion:118]:   Device Model:     NX4832F035_011C
[16:25:09][C][nextion:119]:   Firmware Version: 72
[16:25:09][C][nextion:120]:   Serial Number:    C28C3401E4xxxxxxx
[16:25:09][C][nextion:121]:   Flash Size:       16777216
[16:25:09][C][nextion:122]:   Wake On Touch:    True
[16:25:09][C][ntc:014]: NTC Sensor 'nspanel01 Temperature'
[16:25:09][C][ntc:014]:   Device Class: 'temperature'
[16:25:09][C][ntc:014]:   State Class: 'measurement'
[16:25:09][C][ntc:014]:   Unit of Measurement: '°C'
[16:25:09][C][ntc:014]:   Accuracy Decimals: 1
[16:25:09][C][captive_portal:088]: Captive Portal:
[16:25:09][C][web_server:151]: Web Server:
[16:25:09][C][web_server:152]:   Address: nspanel01.local:80
[16:25:09][C][mdns:108]: mDNS:
[16:25:09][C][mdns:109]:   Hostname: nspanel01
[16:25:09][C][ota:093]: Over-The-Air Updates:
[16:25:09][C][ota:094]:   Address: nspanel01.local:3232
[16:25:09][C][ota:097]:   Using Password.
[16:25:09][C][api:138]: API Server:
[16:25:09][C][api:139]:   Address: nspanel01.local:6053
[16:25:09][C][api:143]:   Using noise encryption: NO
[16:25:09][C][wifi_info:009]: WifiInfo IPAddress 'nspanel01 IP'
[16:25:09][C][wifi_info:011]: WifiInfo SSID 'nspanel01 SSID'
[16:25:09][C][wifi_info:012]: WifiInfo BSSID 'nspanel01 BSSID'
[16:25:09][C][wifi_signal.sensor:009]: WiFi Signal 'nspanel01 RSSI'
[16:25:09][C][wifi_signal.sensor:009]:   Device Class: 'signal_strength'
[16:25:09][C][wifi_signal.sensor:009]:   State Class: 'measurement'
[16:25:09][C][wifi_signal.sensor:009]:   Unit of Measurement: 'dBm'
[16:25:09][C][wifi_signal.sensor:009]:   Accuracy Decimals: 0

**Comment by edwardtfn** [comment]
It looks all good.
Are you sure your ESPHome still v3.x?
Looks like v4.0 was successfully flashed.

**Comment by edwardtfn** [comment]
Let me ask you to try one thing...

On your ESPHome yaml settings, in the remote package area, try changing from `ref: main` to `ref: beta`.
This should force your ESPHome to download the code again, now from beta brach, which at this moment is exactly same code as main.
I just wanna make a test. After that, regardless the results, I would suggest going back to `ref: main`.

**Comment by BananaJon** [comment]
Ok, I will try. 

I change it here and then directly install or first validate?

![image](https://github.com/Blackymas/NSPanel_HA_Blueprint/assets/76597486/107cbf9d-63c1-425c-8bbd-359a083b78fc)

**Comment by edwardtfn** [comment]
Yeah, this is where you change.
Either way if you want to validate first or install directly (which will save you a bit of time).

**Comment by edwardtfn** [comment]
This is similar to #1069

**Comment by BananaJon** [comment]
I changed main to beta, but when I validate or install wireless I get this error:

![image](https://github.com/Blackymas/NSPanel_HA_Blueprint/assets/76597486/9e506113-bdb5-4dda-9cff-6d26c70c85ab)

![image](https://github.com/Blackymas/NSPanel_HA_Blueprint/assets/76597486/d4bb8f3b-b00d-4639-ab0c-1af5dbac858b)

**Comment by edwardtfn** [comment]
Try to clean build files again.

**Comment by edwardtfn** [comment]
Do you know which version is your ESPHome compiler?
It looks like a `dev` build of v2023.4, is that right?
```txt
[16:25:08][I][app:102]: ESPHome version 2023.4.0-dev
```

**Comment by edwardtfn** [comment]
ESPHome v2023.5 is required for v4.0:

https://github.com/Blackymas/NSPanel_HA_Blueprint/blob/main/nspanel_esphome.yaml#L28

**Comment by BananaJon** [comment]
> Try to clean build files again.

Unfortunately does not help. Then it is due to the compiler version

I use ESPHome (dev) in HomeAssistant - HACS and and it should be the latest version available. Automatic updates are active. 

Home Assistant 2023.9.2
Supervisor 2023.09.2
Operating System 10.5
Frontend 20230911.0 - latest

Is it difficult to compile this manually via e.g. Windows and then install it via OTA?

**Comment by BananaJon** [comment]
I have solved the problem. Automatic Updates in HomeAssistant - HACS for ESPHome DEV was active, but apparently it was still not the latest ESPHome DEV version.

I went HACS then to Recreate. 

![image](https://github.com/Blackymas/NSPanel_HA_Blueprint/assets/76597486/3ee1d86a-509c-49e8-aacb-e72d037082bd)

I then validated for the NSPanel and was then able to install the latest ESPHome 4.0.  

At this time the RemotePackage was still on Beta. I have now changed it back to Main.
